求助seata 1.6.0错误问题

io.seata.core.exception.BranchTransactionException: Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681029 branchId = 4323843421687681030 not found service provider for : io.seata.sqlparser.KeywordChecker
	at io.seata.rm.datasource.undo.AbstractUndoLogManager.undo(AbstractUndoLogManager.java:368) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.datasource.DataSourceManager.branchRollback(DataSourceManager.java:122) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.doBranchRollback(AbstractRMHandler.java:125) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler$2.execute(AbstractRMHandler.java:67) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler$2.execute(AbstractRMHandler.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:131) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.handle(AbstractRMHandler.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.DefaultRMHandler.handle(DefaultRMHandler.java:68) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.protocol.transaction.BranchRollbackRequest.handle(BranchRollbackRequest.java:35) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.onRequest(AbstractRMHandler.java:150) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.processor.client.RmBranchRollbackProcessor.handleBranchRollback(RmBranchRollbackProcessor.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.processor.client.RmBranchRollbackProcessor.process(RmBranchRollbackProcessor.java:58) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:281) ~[seata-all-1.6.0.jar:1.6.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.72.Final.jar:4.1.72.Final]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: io.seata.common.loader.EnhancedServiceNotFoundException: not found service provider for : io.seata.sqlparser.KeywordChecker

2023-05-03 20:06:05.560  INFO 33160 --- [_RMROLE_1_19_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:05.560  INFO 33160 --- [_RMROLE_1_21_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681031], branchId:[4323843421687681032], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681031 branchId = 4323843421687681032 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:05.560  INFO 33160 --- [_RMROLE_1_21_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:05.602  INFO 33160 --- [_RMROLE_1_18_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681027], branchId:[4323843421687681028], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681027 branchId = 4323843421687681028 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:05.602  INFO 33160 --- [_RMROLE_1_18_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_23_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681031,branchId=4323843421687681032,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_23_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681031 4323843421687681032 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_20_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681027,branchId=4323843421687681028,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_22_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681029,branchId=4323843421687681030,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_20_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681027 4323843421687681028 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:06.374  INFO 33160 --- [_RMROLE_1_22_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681029 4323843421687681030 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:06.576  INFO 33160 --- [_RMROLE_1_23_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681031], branchId:[4323843421687681032], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681031 branchId = 4323843421687681032 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:06.576  INFO 33160 --- [_RMROLE_1_23_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:06.576  INFO 33160 --- [_RMROLE_1_20_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681027], branchId:[4323843421687681028], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681027 branchId = 4323843421687681028 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:06.576  INFO 33160 --- [_RMROLE_1_20_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:06.651  INFO 33160 --- [_RMROLE_1_22_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681029], branchId:[4323843421687681030], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681029 branchId = 4323843421687681030 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:06.651  INFO 33160 --- [_RMROLE_1_22_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:07.366  INFO 33160 --- [h_RMROLE_1_1_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681029,branchId=4323843421687681030,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:07.366  INFO 33160 --- [h_RMROLE_1_1_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681029 4323843421687681030 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:07.366  INFO 33160 --- [_RMROLE_1_24_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681027,branchId=4323843421687681028,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:07.366  INFO 33160 --- [h_RMROLE_1_2_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681031,branchId=4323843421687681032,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:07.366  INFO 33160 --- [_RMROLE_1_24_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681027 4323843421687681028 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:07.366  INFO 33160 --- [h_RMROLE_1_2_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681031 4323843421687681032 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:07.567  INFO 33160 --- [h_RMROLE_1_1_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681029], branchId:[4323843421687681030], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681029 branchId = 4323843421687681030 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:07.567  INFO 33160 --- [h_RMROLE_1_1_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:07.568  INFO 33160 --- [_RMROLE_1_24_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681027], branchId:[4323843421687681028], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681027 branchId = 4323843421687681028 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:07.568  INFO 33160 --- [_RMROLE_1_24_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:07.627  INFO 33160 --- [h_RMROLE_1_2_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681031], branchId:[4323843421687681032], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681031 branchId = 4323843421687681032 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:07.627  INFO 33160 --- [h_RMROLE_1_2_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_4_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681027,branchId=4323843421687681028,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_4_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681027 4323843421687681028 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_3_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681029,branchId=4323843421687681030,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_5_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681031,branchId=4323843421687681032,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_3_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681029 4323843421687681030 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:08.376  INFO 33160 --- [h_RMROLE_1_5_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681031 4323843421687681032 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:08.591  INFO 33160 --- [h_RMROLE_1_3_24] i.seata.rm.datasource.DataSourceManager  : [stacktrace]branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681029], branchId:[4323843421687681030], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681029 branchId = 4323843421687681030 not found service provider for : io.seata.sqlparser.KeywordChecker]

io.seata.core.exception.BranchTransactionException: Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681029 branchId = 4323843421687681030 not found service provider for : io.seata.sqlparser.KeywordChecker
	at io.seata.rm.datasource.undo.AbstractUndoLogManager.undo(AbstractUndoLogManager.java:368) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.datasource.DataSourceManager.branchRollback(DataSourceManager.java:122) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.doBranchRollback(AbstractRMHandler.java:125) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler$2.execute(AbstractRMHandler.java:67) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler$2.execute(AbstractRMHandler.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:131) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.handle(AbstractRMHandler.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.DefaultRMHandler.handle(DefaultRMHandler.java:68) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.protocol.transaction.BranchRollbackRequest.handle(BranchRollbackRequest.java:35) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.rm.AbstractRMHandler.onRequest(AbstractRMHandler.java:150) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.processor.client.RmBranchRollbackProcessor.handleBranchRollback(RmBranchRollbackProcessor.java:63) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.processor.client.RmBranchRollbackProcessor.process(RmBranchRollbackProcessor.java:58) ~[seata-all-1.6.0.jar:1.6.0]
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:281) ~[seata-all-1.6.0.jar:1.6.0]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.72.Final.jar:4.1.72.Final]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: io.seata.common.loader.EnhancedServiceNotFoundException: not found service provider for : io.seata.sqlparser.KeywordChecker

2023-05-03 20:06:08.591  INFO 33160 --- [h_RMROLE_1_3_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:08.789  INFO 33160 --- [h_RMROLE_1_5_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681031], branchId:[4323843421687681032], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681031 branchId = 4323843421687681032 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:08.789  INFO 33160 --- [h_RMROLE_1_5_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:08.789  INFO 33160 --- [h_RMROLE_1_4_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681027], branchId:[4323843421687681028], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681027 branchId = 4323843421687681028 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:08.789  INFO 33160 --- [h_RMROLE_1_4_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_8_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681031,branchId=4323843421687681032,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_8_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681031 4323843421687681032 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_7_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681029,branchId=4323843421687681030,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_6_24] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=192.168.1.15:8091:4323843421687681027,branchId=4323843421687681028,branchType=AT,resourceId=jdbc:mariadb://124.70.100.65:63706/rocketsecond,applicationData={"autoCommit":false,"skipCheckLock":true}
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_7_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681029 4323843421687681030 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:09.365  INFO 33160 --- [h_RMROLE_1_6_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 192.168.1.15:8091:4323843421687681027 4323843421687681028 jdbc:mariadb://124.70.100.65:63706/rocketsecond
2023-05-03 20:06:09.538  INFO 33160 --- [h_RMROLE_1_8_24] i.seata.rm.datasource.DataSourceManager  : branchRollback failed. branchType:[AT], xid:[192.168.1.15:8091:4323843421687681031], branchId:[4323843421687681032], resourceId:[jdbc:mariadb://124.70.100.65:63706/rocketsecond], applicationData:[{"autoCommit":false,"skipCheckLock":true}]. reason:[Branch session rollback failed and try again later xid = 192.168.1.15:8091:4323843421687681031 branchId = 4323843421687681032 not found service provider for : io.seata.sqlparser.KeywordChecker]
2023-05-03 20:06:09.538  INFO 33160 --- [h_RMROLE_1_8_24] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_RollbackFailed_Retryable

--- a/changes/en-us/develop.md
+++ b/changes/en-us/develop.md
@@ -33,6 +33,7 @@ Add changes here for all PR submitted to the develop branch.
 - [[#5538](https://github.com/seata/seata/pull/5538)] fix finished transaction swallows exception when committing
 - [[#5539](https://github.com/seata/seata/pull/5539)] fix the full table scan issue with 'setDate' condition in Oracle 10g
 - [[#5540](https://github.com/seata/seata/pull/5540)] fix GlobalStatus=9 can't be cleared in DB storage mode
+- [[#5552](https://github.com/seata/seata/pull/5552)] fix mariadb rollback failed
 
 ### optimize:
 - [[#5208](https://github.com/seata/seata/pull/5208)] optimize throwable getCause once more

--- a/changes/zh-cn/develop.md
+++ b/changes/zh-cn/develop.md
@@ -33,6 +33,7 @@
 - [[#5538](https://github.com/seata/seata/pull/5538)] 修复提交事务时事务已完成不抛出异常问题
 - [[#5539](https://github.com/seata/seata/pull/5539)] 修复Oracle 10g where条件包含setDate全表扫描问题
 - [[#5540](https://github.com/seata/seata/pull/5540)] 修复 GlobalStatus=9 在DB存储模式无法清除的问题
+- [[#5552](https://github.com/seata/seata/pull/5552)] 修复mariadb回滚失败的问题
 
 ### optimize:
 - [[#5208](https://github.com/seata/seata/pull/5208)] 优化多次重复获取Throwable#getCause问题

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
@@ -34,6 +34,7 @@ import io.seata.common.util.StringUtils;
 import io.seata.config.ConfigurationFactory;
 import io.seata.core.constants.ConfigurationKeys;
 import io.seata.core.model.Result;
+import io.seata.rm.datasource.ConnectionProxy;
 import io.seata.sqlparser.util.ColumnUtils;
 import io.seata.rm.datasource.DataCompareUtils;
 import io.seata.rm.datasource.SqlGenerateUtils;
@@ -110,11 +111,12 @@ public abstract class AbstractUndoExecutor {
     /**
      * Execute on.
      *
-     * @param conn the conn
+     * @param connectionProxy the connection proxy
      * @throws SQLException the sql exception
      */
-    public void executeOn(Connection conn) throws SQLException {
-        if (IS_UNDO_DATA_VALIDATION_ENABLE && !dataValidationAndGoOn(conn)) {
+    public void executeOn(ConnectionProxy connectionProxy) throws SQLException {
+        Connection conn = connectionProxy.getTargetConnection();
+        if (IS_UNDO_DATA_VALIDATION_ENABLE && !dataValidationAndGoOn(connectionProxy)) {
             return;
         }
         PreparedStatement undoPST = null;
@@ -228,7 +230,7 @@ public abstract class AbstractUndoExecutor {
      * @return return true if data validation is ok and need continue undo, and return false if no need continue undo.
      * @throws SQLException the sql exception such as has dirty data
      */
-    protected boolean dataValidationAndGoOn(Connection conn) throws SQLException {
+    protected boolean dataValidationAndGoOn(ConnectionProxy conn) throws SQLException {
 
         TableRecords beforeRecords = sqlUndoLog.getBeforeImage();
         TableRecords afterRecords = sqlUndoLog.getAfterImage();
@@ -251,7 +253,7 @@ public abstract class AbstractUndoExecutor {
         Result<Boolean> afterEqualsCurrentResult = DataCompareUtils.isRecordsEquals(afterRecords, currentRecords);
         if (!afterEqualsCurrentResult.getResult()) {
 
-            // If current data is not equivalent to the after data, then compare the current data with the before 
+            // If current data is not equivalent to the after data, then compare the current data with the before
             // data, too. No need continue to undo if current data is equivalent to the before data snapshot
             Result<Boolean> beforeEqualsCurrentResult = DataCompareUtils.isRecordsEquals(beforeRecords, currentRecords);
             if (beforeEqualsCurrentResult.getResult()) {
@@ -282,11 +284,12 @@ public abstract class AbstractUndoExecutor {
     /**
      * Query current records.
      *
-     * @param conn the conn
+     * @param connectionProxy the connection proxy
      * @return the table records
      * @throws SQLException the sql exception
      */
-    protected TableRecords queryCurrentRecords(Connection conn) throws SQLException {
+    protected TableRecords queryCurrentRecords(ConnectionProxy connectionProxy) throws SQLException {
+        Connection conn = connectionProxy.getTargetConnection();
         TableRecords undoRecords = getUndoRows();
         TableMeta tableMeta = undoRecords.getTableMeta();
         //the order of element matters
@@ -301,7 +304,7 @@ public abstract class AbstractUndoExecutor {
         String firstKey = pkRowValues.keySet().stream().findFirst().get();
         int pkRowSize = pkRowValues.get(firstKey).size();
         String checkSQL = String.format(CHECK_SQL_TEMPLATE, sqlUndoLog.getTableName(),
-                SqlGenerateUtils.buildWhereConditionByPKs(pkNameList, pkRowSize, getDbType(conn)));
+                SqlGenerateUtils.buildWhereConditionByPKs(pkNameList, pkRowSize, connectionProxy.getDbType()));
 
         PreparedStatement statement = null;
         ResultSet checkSet = null;

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
@@ -44,7 +44,6 @@ import io.seata.rm.datasource.sql.struct.KeyType;
 import io.seata.rm.datasource.sql.struct.Row;
 import io.seata.rm.datasource.sql.struct.TableMeta;
 import io.seata.rm.datasource.sql.struct.TableRecords;
-import io.seata.rm.datasource.util.JdbcUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java
@@ -126,7 +126,7 @@ public abstract class AbstractUndoExecutor {
             TableRecords undoRows = getUndoRows();
             for (Row undoRow : undoRows.getRows()) {
                 ArrayList<Field> undoValues = new ArrayList<>();
-                List<Field> pkValueList = getOrderedPkList(undoRows, undoRow, getDbType(conn));
+                List<Field> pkValueList = getOrderedPkList(undoRows, undoRow, connectionProxy.getDbType());
                 for (Field field : undoRow.getFields()) {
                     if (field.getKeyType() != KeyType.PRIMARY_KEY) {
                         undoValues.add(field);
@@ -381,17 +381,6 @@ public abstract class AbstractUndoExecutor {
         }
         Map<String, List<Field>> pkValueMap = pkFieldList.stream().collect(Collectors.groupingBy(Field::getName));
         return pkValueMap;
-    }
-
-    /**
-     * Get db type
-     *
-     * @param conn the connection
-     * @return the db type
-     * @throws SQLException SQLException
-     */
-    protected String getDbType(Connection conn) throws SQLException {
-        return JdbcUtils.getDbType(conn.getMetaData().getURL());
     }
 
 }

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
@@ -255,6 +255,7 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
      */
     @Override
     public void undo(DataSourceProxy dataSourceProxy, String xid, long branchId) throws TransactionException {
+        ConnectionProxy connectionProxy = null;
         Connection conn = null;
         ResultSet rs = null;
         PreparedStatement selectPST = null;
@@ -262,8 +263,8 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
 
         for (; ; ) {
             try {
-                conn = dataSourceProxy.getPlainConnection();
-
+                connectionProxy = dataSourceProxy.getConnection();
+                conn = connectionProxy.getTargetConnection();
                 // The entire undo process should run in a local transaction.
                 if (originalAutoCommit = conn.getAutoCommit()) {
                     conn.setAutoCommit(false);
@@ -312,7 +313,7 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
                             sqlUndoLog.setTableMeta(tableMeta);
                             AbstractUndoExecutor undoExecutor = UndoExecutorFactory.getUndoExecutor(
                                 dataSourceProxy.getDbType(), sqlUndoLog);
-                            undoExecutor.executeOn(conn);
+                            undoExecutor.executeOn(connectionProxy);
                         }
                     } finally {
                         // remove serializer name
@@ -381,7 +382,7 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
                         if (originalAutoCommit) {
                             conn.setAutoCommit(true);
                         }
-                        conn.close();
+                        connectionProxy.close();
                     }
                 } catch (SQLException closeEx) {
                     LOGGER.warn("Failed to close JDBC resource while undo ... ", closeEx);

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
@@ -265,6 +265,7 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
             try {
                 connectionProxy = dataSourceProxy.getConnection();
                 conn = connectionProxy.getTargetConnection();
+				
                 // The entire undo process should run in a local transaction.
                 if (originalAutoCommit = conn.getAutoCommit()) {
                     conn.setAutoCommit(false);

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java
@@ -265,7 +265,7 @@ public abstract class AbstractUndoLogManager implements UndoLogManager {
             try {
                 connectionProxy = dataSourceProxy.getConnection();
                 conn = connectionProxy.getTargetConnection();
-				
+
                 // The entire undo process should run in a local transaction.
                 if (originalAutoCommit = conn.getAutoCommit()) {
                     conn.setAutoCommit(false);

--- a/rm-datasource/src/test/java/io/seata/rm/datasource/undo/BaseH2Test.java
+++ b/rm-datasource/src/test/java/io/seata/rm/datasource/undo/BaseH2Test.java
@@ -16,6 +16,8 @@
 package io.seata.rm.datasource.undo;
 
 import io.seata.common.util.IOUtil;
+import io.seata.rm.datasource.ConnectionProxy;
+import io.seata.rm.datasource.DataSourceProxy;
 import io.seata.rm.datasource.sql.struct.ColumnMeta;
 import io.seata.rm.datasource.sql.struct.Field;
 import io.seata.rm.datasource.sql.struct.KeyType;
@@ -44,7 +46,9 @@ public abstract class BaseH2Test {
     
     static BasicDataSource dataSource = null;
 
-    static Connection connection = null;
+    static ConnectionProxy connection = null;
+
+    static DataSourceProxy dataSourceProxy = null;
 
     static TableMeta tableMeta = null;
     
@@ -55,8 +59,8 @@ public abstract class BaseH2Test {
         dataSource.setUrl("jdbc:h2:./db_store/test_undo");
         dataSource.setUsername("sa");
         dataSource.setPassword("");
-
-        connection = dataSource.getConnection();
+        dataSourceProxy = new DataSourceProxy(dataSource);
+        connection = dataSourceProxy.getConnection();
 
         tableMeta = mockTableMeta();
     }

--- a/rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java
+++ b/rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java
@@ -49,6 +49,9 @@ import java.util.concurrent.Executor;
 import com.alibaba.fastjson.JSON;
 import com.alibaba.fastjson.serializer.SerializerFeature;
 
+import io.seata.rm.datasource.ConnectionProxy;
+import io.seata.rm.datasource.DataSourceProxy;
+import io.seata.rm.datasource.mock.MockDataSource;
 import io.seata.sqlparser.SQLType;
 import io.seata.rm.datasource.sql.struct.Field;
 import io.seata.rm.datasource.sql.struct.KeyType;
@@ -56,6 +59,7 @@ import io.seata.rm.datasource.sql.struct.Row;
 import io.seata.rm.datasource.sql.struct.TableMeta;
 import io.seata.rm.datasource.sql.struct.TableRecords;
 import io.seata.sqlparser.util.JdbcConstants;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.mockito.Mockito;
 
@@ -63,6 +67,12 @@ import org.mockito.Mockito;
  * The type Undo executor test.
  */
 public class UndoExecutorTest {
+
+    MockConnection connection = new MockConnection();
+    MockDataSource dataSource = new MockDataSource();
+    DataSourceProxy dataSourceProxy = new DataSourceProxy(dataSource);
+    ConnectionProxy connectionProxy = new ConnectionProxy(dataSourceProxy, connection);
+
 
     /**
      * Test field.
@@ -146,12 +156,11 @@ public class UndoExecutorTest {
         SQLUndoLog.setAfterImage(afterImage);
 
         AbstractUndoExecutor executor = UndoExecutorFactory.getUndoExecutor(JdbcConstants.MYSQL, SQLUndoLog);
-        MockConnection connection = new MockConnection();
         AbstractUndoExecutor spy = Mockito.spy(executor);
         // skip data validation
-        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connection);
-        Mockito.doReturn(JdbcConstants.MYSQL).when(spy).getDbType(connection);
-        spy.executeOn(connection);
+        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connectionProxy);
+        Assertions.assertEquals(JdbcConstants.MYSQL,connectionProxy.getDbType());
+        spy.executeOn(connectionProxy);
     }
 
     /**
@@ -216,12 +225,11 @@ public class UndoExecutorTest {
         SQLUndoLog.setAfterImage(afterImage);
 
         AbstractUndoExecutor executor = UndoExecutorFactory.getUndoExecutor(JdbcConstants.MYSQL, SQLUndoLog);
-        MockConnection connection = new MockConnection();
         AbstractUndoExecutor spy = Mockito.spy(executor);
         // skip data validation
-        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connection);
-        Mockito.doReturn(JdbcConstants.MYSQL).when(spy).getDbType(connection);
-        spy.executeOn(connection);
+        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connectionProxy);
+        Assertions.assertEquals(JdbcConstants.MYSQL,connectionProxy.getDbType());
+        spy.executeOn(connectionProxy);
     }
 
     /**
@@ -286,12 +294,12 @@ public class UndoExecutorTest {
         SQLUndoLog.setBeforeImage(beforeImage);
 
         AbstractUndoExecutor executor = UndoExecutorFactory.getUndoExecutor(JdbcConstants.MYSQL, SQLUndoLog);
-        MockConnection connection = new MockConnection();
+
         AbstractUndoExecutor spy = Mockito.spy(executor);
         // skip data validation
-        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connection);
-        Mockito.doReturn(JdbcConstants.MYSQL).when(spy).getDbType(connection);
-        spy.executeOn(connection);
+        Mockito.doReturn(true).when(spy).dataValidationAndGoOn(connectionProxy);
+        Assertions.assertEquals(JdbcConstants.MYSQL,connectionProxy.getDbType());
+        spy.executeOn(connectionProxy);
     }
 
     /**

